### PR TITLE
Add uart support for nxp_adsp_8m

### PR DESCRIPTION
**Pull request by iuliana-prodan**
In order to enable, for audio DSP, the UART console (for now) and use the driver from NXP hal I needed to rename SOC and add uart node to dts.

I also renamed the  SoC overlay for NXP i.MX8MP from nxp_imx8m to mimx8ml8, in xtensa_hal - see https://github.com/zephyrproject-rtos/hal_xtensa/pull/18

### CODE REVIEW

**Comment by iuliana-prodan** [comment]
I set this to draft, until the xtensa_hal PR is merged and mcu-sdk from nxp_hal is updated.
When all these are done, will add a new commit to this PR - to update west.yaml with new revison id - and have all tests pass. 

cc: @nashif @dleach02 @dbaluta

**Comment by iuliana-prodan** [comment]
@nashif @dleach02 I added a commit (to update west.yaml - see https://github.com/zephyrproject-rtos/zephyr/pull/57084/commits/2cc51ac29bd073d1662c44cfcb3e217bc6521a7c) to test all modification with the new mcu-sdk (from nxp_hal) and xtensa_hal (https://github.com/zephyrproject-rtos/hal_xtensa/pull/18) and everything seems good.

One problem, though - when updating xtensa hal (is also included commit https://github.com/zephyrproject-rtos/hal_xtensa/commit/e636ce8fdf334addaae37fb2417dc2cb8bf7b50a) witch seems to break intel_adsp_ace15_mtpm. This commit is not included in the xtensa_hal used _now_ in zephyr - the revision used is https://github.com/zephyrproject-rtos/hal_xtensa/commit/63f655362423aa49507da7977a2d37142e8debeb.

@nashif So, please, we either revert https://github.com/zephyrproject-rtos/hal_xtensa/commit/e636ce8fdf334addaae37fb2417dc2cb8bf7b50a since is not used, or.. how do you see this fixed?

Thanks!

cc: @dcpleung

**Comment by dcpleung** [comment]
That change for intel_adsp_ace15_mtpm needs a new toolchain to work. So we are waiting for a new Zephyr SDK before moving to the new SHA. I think we can revert that in the meantime while we are working on this, and re-introduce it later.

**Comment by iuliana-prodan** [comment]
> @nashif @dleach02 I added a commit (to update west.yaml - see [2cc51ac](https://github.com/zephyrproject-rtos/zephyr/commit/2cc51ac29bd073d1662c44cfcb3e217bc6521a7c)) to test all modification with the new mcu-sdk (from nxp_hal) and xtensa_hal ([zephyrproject-rtos/hal_xtensa#18](https://github.com/zephyrproject-rtos/hal_xtensa/pull/18)) and everything seems good.
> 
> One problem, though - when updating xtensa hal (is also included commit [zephyrproject-rtos/hal_xtensa@e636ce8](https://github.com/zephyrproject-rtos/hal_xtensa/commit/e636ce8fdf334addaae37fb2417dc2cb8bf7b50a)) witch seems to break intel_adsp_ace15_mtpm. This commit is not included in the xtensa_hal used _now_ in zephyr - the revision used is [zephyrproject-rtos/hal_xtensa@63f6553](https://github.com/zephyrproject-rtos/hal_xtensa/commit/63f655362423aa49507da7977a2d37142e8debeb).
> 
> @nashif So, please, we either revert [zephyrproject-rtos/hal_xtensa@e636ce8](https://github.com/zephyrproject-rtos/hal_xtensa/commit/e636ce8fdf334addaae37fb2417dc2cb8bf7b50a) since is not used, or.. how do you see this fixed?
> 
> Thanks!
> 
> cc: @dcpleung

@nashif @dcpleung only the xtensa_hal needs to be updated (the nxp_hal has been updated in another PR).
Please, let me know how do you want me to continue with this PR and the xtensa hal: https://github.com/zephyrproject-rtos/hal_xtensa/pull/18

**Comment by iuliana-prodan** [comment]
> That change for intel_adsp_ace15_mtpm needs a new toolchain to work. So we are waiting for a new Zephyr SDK before moving to the new SHA. I think we can revert that in the meantime while we are working on this, and re-introduce it later.

@dcpleung can I revert that commit?
And add the revert commit in this PR: https://github.com/zephyrproject-rtos/hal_xtensa/pull/18, so we can test the new xtensa_hal revision?

**Comment by dbaluta** [comment]
@iuliana-prodan lets go with revert as @dcpleung suggested in a comment above.

**Comment by iuliana-prodan** [comment]
> @iuliana-prodan lets go with revert as @dcpleung suggested in a comment above.

Done!
I've updated PR: https://github.com/zephyrproject-rtos/hal_xtensa/pull/18/commits

**Comment by iuliana-prodan** [comment]
In order to compile an application and enable UART a user must run west build using `DTC_OVERLAY_FILE` and `CONF_FILE`.
Here's an example for `hello_world`:
`west build -p always -b nxp_adsp_imx8m samples/hello_world/ -DDTC_OVERLAY_FILE="boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m_uart.overlay" -DCONF_FILE="boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m_uart.conf"`

**Comment by iuliana-prodan** [comment]
@dcpleung @nashif @dleach02 can you, please, have a look?
All tests are passing - but we first need to merge: https://github.com/zephyrproject-rtos/hal_xtensa/pull/18

Thanks!

**Comment by iuliana-prodan** [comment]
> @iuliana-prodan Please do not remove the DNM label added by the bot. It is added due to your change to west.yml, pulling in a PR head instead of a SHA.
> 
> Once [zephyrproject-rtos/hal_xtensa#18](https://github.com/zephyrproject-rtos/hal_xtensa/pull/18) is merged, you need to update the commit to west.yml to pull in the SHA instead - and then the bot will automatically remove the DNM label.

Thanks @henrikbrixandersen! I didn't know it can be automatically removed.

**Review by dleach02 — DISMISSED** [review]

**Review by dbaluta — DISMISSED** [review]

**Review by LaurentiuM1234 — COMMENTED** [review]
A few questions/suggestions, otherwise LGTM.

File: dts/xtensa/nxp/nxp_imx8m.dtsi [hunk 1]
```diff
@@ -30,4 +30,34 @@
...
+		};
+
+		/*
+		 * For now only UART4 is supported and
+		 * tested with the serial driver
+		 */
+		uart4: uart@30a60000 {
+			compatible = "nxp,imx-iuart";
+			reg = <0x30a60000 0x10000>;
+			//interrupts = <29 3>;
```
[the remark this answers is not in the document]
Comment (iuliana-prodan) [reply dts/xtensa/nxp/nxp_imx8m.dtsi]: yes, I removed it, for now.

File: dts/xtensa/nxp/nxp_imx8m.dtsi [hunk 2]
```diff
@@ -30,4 +30,34 @@
...
+			clocks = <&ccm 0x0803 0x6c 24>;
```
[the remark this answers is not in the document]
Comment (iuliana-prodan) [reply dts/xtensa/nxp/nxp_imx8m.dtsi]: done.

**Review by dbaluta — APPROVED** [review]

**Review by LaurentiuM1234 — APPROVED** [review]

**Review by dleach02 — APPROVED** [review]

**Review by danieldegrasse — APPROVED** [review]

**Comment by marc-hb** [comment]
@iuliana-prodan would it be possible to update `hal_xtensa` in `sof/west.yml` too?

I don't understand all the details sorry, I only spotted this:

https://github.com/thesofproject/sof/actions/runs/4927775972/jobs/8805235896?pr=7565
```
-- Found host-tools: zephyr 0.15.2 (/home/cwd_user/zephyr-sdk-0.15.2)
-- Found toolchain: zephyr 0.15.2 (/home/cwd_user/zephyr-sdk-0.15.2)
-- Found Dtc: /home/cwd_user/zephyr-sdk-0.15.2/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6") 
-- Found BOARD.dts: /zep_workspace/zephyr/boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m.dts
In file included from <command-line>:
/zep_workspace/zephyr/boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m.dts:10:10: fatal error: nxp/nxp_imx/mimx8ml8dvnlz-pinctrl.dtsi: No such file or directory
   10 | #include <nxp/nxp_imx/mimx8ml8dvnlz-pinctrl.dtsi>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by iuliana-prodan** [comment]
> @iuliana-prodan would it be possible to update `hal_xtensa` in `sof/west.yml` too?
> 
> I don't understand all the details sorry, I only spotted this:
> 
> https://github.com/thesofproject/sof/actions/runs/4927775972/jobs/8805235896?pr=7565
> 
> ```
> -- Found host-tools: zephyr 0.15.2 (/home/cwd_user/zephyr-sdk-0.15.2)
> -- Found toolchain: zephyr 0.15.2 (/home/cwd_user/zephyr-sdk-0.15.2)
> -- Found Dtc: /home/cwd_user/zephyr-sdk-0.15.2/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6") 
> -- Found BOARD.dts: /zep_workspace/zephyr/boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m.dts
> In file included from <command-line>:
> /zep_workspace/zephyr/boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m.dts:10:10: fatal error: nxp/nxp_imx/mimx8ml8dvnlz-pinctrl.dtsi: No such file or directory
>    10 | #include <nxp/nxp_imx/mimx8ml8dvnlz-pinctrl.dtsi>
>       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```

@marc-hb  There is no need to update hal_xtensa in [sof/west.yml](https://github.com/thesofproject/sof/blob/main/west.yml#L56). 
We are not using a specific one, this is imported - it uses the one from zephyr's west.yml.
So, when updating zephyr revision, we should do a ` west update` to take, also, any modifications done on west.yml.

**Comment by marc-hb** [comment]
> We are not using a specific one, this is imported - it uses the one from zephyr's west.yml.

I forgot this, apologies. Thanks for the reminder and clarification.

**Comment by marc-hb** [comment]
> So, when updating zephyr revision, we should do a  west update to take, also, any modifications done on west.yml.

So now I tried that and it still failed the same as above.

Next I put SOF CI totally aside and tried this above:

> west build -p always -b nxp_adsp_imx8m samples/hello_world/ -DDTC_OVERLAY_FILE="boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m_uart.overlay" -DCONF_FILE="boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m_uart.conf"

It failed with multiple -Werrors below.

The previous `west build` command failed too in some other way. Long story short I could not get anything to compile cleanly with a vanilla Zephyr tree.


@iuliana-prodan can you please try from scratch from a really a pristine tree and try these again?

```
modules/audio/sof/src/arch/xtensa/hal/atomics.c:10:38: error: passing argument 1 of 'xthal_compare_and_set' from incompatible pointer type [-Werror=incompatible-pointer-types]
   10 |         return xthal_compare_and_set(address, test_value, set_value);
      |                                      ^~~~~~~
      |                                      |
      |                                      int32_t * {aka long int *}
In file included from modules/audio/sof/src/arch/xtensa/hal/atomics.c:4:
modules/audio/sof/src/arch/xtensa/include/xtensa/hal.h:387:45: note: expected 'int *' but argument is of type 'int32_t *' {aka 'long int *'}
  387 | extern int      xthal_compare_and_set( int *addr, int test_val, int compare_val );
      |                                        ~~~~~^~~~
cc1: all warnings being treated as errors
[61/212] Building C object src/init/CMakeFiles/ext_manifest.dir/ext_manifest.c.o
FAILED: src/init/CMakeFiles/ext_manifest.dir/ext_manifest.c.o 
/var/home/mherber2/zephyr-sdk-0.15.2/xtensa-nxp_imx8m_adsp_zephyr-elf/bin/xtensa-nxp_imx8m_adsp_zephyr-elf-gcc -DBLD_COUNTERS=0 -DCC_OPTIMIZE_FLAGS=\"-O2\" -DXCC_TOOLS_VERSION=\"xtensa-nxp_imx8m_adsp_zephyr-elf\" -DRELATIVE_FILE=\"src/init/ext_manifest.c\" -Imodules/audio/sof/src/arch/xtensa/include -Imodules/audio/sof/src/arch/xtensa/xtos -Imodules/audio/sof/xtos/include -Imodules/audio/sof/src/platform/imx8m/include -Imodules/audio/sof/src/platform/imx8m/include/arch -I/include -Imodules/audio/sof/src/include -Izephyr/build/generated/include -Imodules/audio/sof/third_party_include -nostdlib -fno-inline-functions -mlongcalls -O2 -g -Wall -Werror -Wl,-EL -Wmissing-prototypes -Wpointer-arith -mtext-section-literals -imacroszephyr/build/generated/include/autoconfig.h -MD -MT src/init/CMakeFiles/ext_manifest.dir/ext_manifest.c.o -MF src/init/CMakeFiles/ext_manifest.dir/ext_manifest.c.o.d -o src/init/CMakeFiles/ext_manifest.dir/ext_manifest.c.o -c modules/audio/sof/src/init/ext_manifest.c
In file included from modules/audio/sof/xtos/include/sof/compiler_info.h:17,
                 from modules/audio/sof/src/init/ext_manifest.c:10:
modules/audio/sof/src/arch/xtensa/include/arch/compiler_info.h:40:17: error: initializer-string for array of 'unsigned char' is too long [-Werror]
   40 | #define CC_DESC " " XCC_TOOLS_VERSION
      |                 ^~~
modules/audio/sof/src/init/ext_manifest.c:59:25: note: in expansion of macro 'CC_DESC'
   59 |                 .desc = CC_DESC "\0", ///< eg. " RG-2017.8-linux"
      |                         ^~~~~~~
modules/audio/sof/src/arch/xtensa/include/arch/compiler_info.h:40:17: note: (near initialization for 'ext_man_cc_ver.cc_version.desc')
   40 | #define CC_DESC " " XCC_TOOLS_VERSION
      |                 ^~~
modules/audio/sof/src/init/ext_manifest.c:59:25: note: in expansion of macro 'CC_DESC'
   59 |                 .desc = CC_DESC "\0", ///< eg. " RG-2017.8-linux"
      |                         ^~~~~~~
cc1: all warnings being treated as errors

```

**Comment by iuliana-prodan** [comment]
> xtensa-nxp_imx8m_adsp_zephyr

Answered here: https://github.com/thesofproject/sof/pull/7586#issuecomment-1543796324

**Pull request merged**